File: CPP/algorithms/graph_algorithms/floyd_warshall.cpp
// Floyd–Warshall Algorithm — All-Pairs Shortest Path
// Problem
//   Compute shortest distances between every pair of vertices in a weighted graph.
//
// Approach
//   1. Initialize dist[i][j]:
//        - 0 if i==j
//        - weight(u,v) if edge exists
//        - INF otherwise
//   2. For k in [0..V-1] (intermediate vertex):
//        for i in [0..V-1] (source):
//          for j in [0..V-1] (destination):
//             dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
//   3. After all iterations, dist[i][j] is the shortest distance.
//
// Complexity
//   Time  : O(V^3)
//   Space : O(V^2)

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int V, E;
  cin >> V >> E;
  const int INF = 1e9;

  vector<vector<int>> dist(V, vector<int>(V, INF));
  for (int i = 0; i < V; i++)
    dist[i][i] = 0;

  for (int i = 0; i < E; i++)
  {
    int u, v, w;
    cin >> u >> v >> w;
    dist[u][v] = min(dist[u][v], w);
    dist[v][u] = min(dist[v][u], w); // remove if directed
  }

  cout << "Initial distance matrix:\n";
  for (auto &row : dist)
  {
    for (auto x : row)
      cout << (x == INF ? -1 : x) << " ";
    cout << "\n";
  }
  cout << "\n";

  // Floyd–Warshall dry run
  for (int k = 0; k < V; k++)
  {
    cout << "Considering intermediate vertex k=" << k << ":\n";
    for (int i = 0; i < V; i++)
      for (int j = 0; j < V; j++)
        if (dist[i][k] < INF && dist[k][j] < INF)
          dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);

    // print matrix after this k
    for (auto &row : dist)
    {
      for (auto x : row)
        cout << (x == INF ? -1 : x) << " ";
      cout << "\n";
    }
    cout << "\n";
  }

  cout << "Final All-Pairs Shortest Distances:\n";
  for (auto &row : dist)
  {
    for (auto x : row)
      cout << (x == INF ? -1 : x) << " ";
    cout << "\n";
  }

  return 0;
}

/*
Example Input:
4 5
0 1 5
0 3 10
1 2 3
2 3 1
0 2 100

Visualization:
Graph:
   (5)        (3)       (1)
0 ----- 1 -------- 2 -------- 3
 \                               ^
  \____________ (10) ___________/

Dry Run Steps:

Initial dist:
0   5 100 10
5   0   3 -1
100 3   0  1
10  -1  1  0

k = 0 (using vertex 0 as intermediate):
- dist[2][3] = min(1, dist[2][0]+dist[0][3]) = min(1,100+10)=1 (no change)
- dist[2][0] = min(100, dist[2][0]+dist[0][0]) = 100 (no change)
... print matrix

k = 1 (vertex 1 intermediate):
- dist[0][2] = min(100, dist[0][1]+dist[1][2]) = min(100,5+3)=8
- dist[3][0] = min(10, dist[3][1]+dist[1][0]) = min(10, INF+5)=10 (no change)
... print matrix

k = 2 (vertex 2 intermediate):
- dist[0][3] = min(10, dist[0][2]+dist[2][3]) = min(10,8+1)=9
- dist[1][3] = min(INF, 3+1)=4
... print matrix

k = 3 (vertex 3 intermediate):
- no further updates

Final dist:
0 5 8 9
5 0 3 4
8 3 0 1
9 4 1 0
*/

// prepare PR

File: Python/data_structures/graphs/floyd_warshall.py
"""
Floyd-Warshall Algorithm
------------------------
Description:
    Computes shortest paths between all pairs of vertices in a weighted graph using dynamic programming.
    Supports negative edge weights but does not detect negative weight cycles.

Algorithm:
    - Initialize a distance matrix with infinity for all pairs except self-loops (0).
    - Populate the matrix with direct edge weights.
    - For each intermediate vertex k, update dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]).

Time Complexity: O(V^3)
Space Complexity: O(V^2)

"""

def floyd_warshall(vertices, edges):
    # Validate that vertex list is not empty
    if not vertices:
        raise ValueError("Vertex list is empty.")

    # Validate that all edge endpoints exist in the graph
    for u, v, _ in edges:
        if u not in vertices or v not in vertices:
            raise ValueError(f"Edge contains undefined vertex: ({u}, {v})")

    # Step 1: Initialize distance matrix with infinity
    dist = {i: {j: float('inf') for j in vertices} for i in vertices}
    for v in vertices:
        dist[v][v] = 0  # Distance to self is zero

    # Step 2: Populate initial edge weights
    for u, v, w in edges:
        dist[u][v] = w

    # Step 3: Floyd-Warshall core logic
    for k in vertices:
        for i in vertices:
            for j in vertices:
                # If going through k offers a shorter path, update it
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    return dist


# ------------------ Test Cases ------------------

def run_tests():
    print("Running Floyd-Warshall test cases...\n")

    # Test Case 1: Basic graph
    vertices1 = ['A', 'B', 'C']
    edges1 = [
        ('A', 'B', 4),
        ('B', 'C', 1),
        ('A', 'C', 10)
    ]
    print("Test Case 1:")
    print(floyd_warshall(vertices1, edges1), "\n")

    # Test Case 2: Graph with negative weights
    vertices2 = ['X', 'Y', 'Z']
    edges2 = [
        ('X', 'Y', 3),
        ('Y', 'Z', -2),
        ('Z', 'X', 5)
    ]
    print("Test Case 2:")
    print(floyd_warshall(vertices2, edges2), "\n")

    # Test Case 3: Disconnected graph
    vertices3 = ['P', 'Q', 'R']
    edges3 = [
        ('P', 'Q', 7)
    ]
    print("Test Case 3:")
    print(floyd_warshall(vertices3, edges3), "\n")

    # Test Case 4: Invalid edge vertex
    try:
        print("Test Case 4:")
        floyd_warshall(['A', 'B'], [('A', 'C', 2)])
    except ValueError as e:
        print("Exception caught -", e, "\n")

    # Test Case 5: Empty vertex list
    try:
        print("Test Case 5:")
        floyd_warshall([], [('A', 'B', 1)])
    except ValueError as e:
        print("Exception caught -", e, "\n")


if __name__ == "__main__":
    run_tests()
